File: 08_JUNE_2021/threeSum.cpp
#include <vector>
#include <iostream>
using namespace std;

void displayVector(vector<vector<int>> input) {
	for (int i = 0; i < input.size(); i++) {
		for (int j = 0; j < input[i].size(); j++) {
			cout << input[i][j] << " ";
		}
		cout << endl;
	}
}

class Solution {
public:
	vector<vector<int>> threeSum(vector<int> numbers, int target);
};

int main(void) {
	Solution solution;
	vector<int> input = {5, 4, 3, 2, 1};
	int target = 5;
	vector<vector<int>> ans = solution.threeSum(input, target);
	displayVector(ans);
}

vector<vector<int>> Solution::threeSum(vector<int> numbers, int target) {
	vector<vector<int>> ans;
	vector<int> inner;

	for (int i = 0; i < numbers.size() - 2; i++) {
		for (int j = i + 1; j < numbers.size() - 1; j++) {
			for (int k = j + 1; k < numbers.size(); k++) {
				if (i + j + k == target) {
					inner = {numbers[i], numbers[j], numbers[k]};
					ans.push_back(inner);
				}	
			}
		}	
	}
	return ans;
}

File: 08_JUNE_2021/threeSum.js
// Find all possible array of 3 unique indices of a 
// given n array that sum of all pointed numbers must be equal
// to a specific target

const helper = (numbers, target, memo ={}) => {
	if (target in memo) return memo[target];
	if (target < 0) return null;
	if (target == 0) return [[numbers[0]]];

	const ans = []
	for (let i = 1; i < numbers.length; i++) {
		const remain = target - i;
		let result = helper(numbers, remain, memo);
		if (result != null) {
			result = result.map(r => [...r, numbers[i]]);
			ans.push(...result);
		}
	}
	memo[target] = ans;
	return ans;
}

const threeSum = (numbers, target) => {
	let ans = helper(numbers, target);
	const memo = {};
	ans = [... new Set(ans.filter(a => a.length == 3).map(a => a.sort().join('')))];
	ans = ans.map(a => a.split('').map(num => +num));
	return ans;
}

console.log(threeSum([5, 4, 3, 2, 1], 5));